File: python/261.py
"""Project Euler Problem 261: Pivotal Square Sums.

Call k a square-pivot if there exist m > 0 and n >= k such that the sum of
the m+1 consecutive squares up to k equals the sum of the m consecutive
squares starting from (n+1). Find the sum of all distinct square pivots up to N.

We rewrite the equation as a general Pell equation x^2 - D*y^2 = N_pell,
with D = m*(m+1), N_pell = m^2*(m+1).

The fundamental solution to x^2 - D*y^2 = 1 is always (2m+1, 2).
Base solutions have y <= m and y^2 divisible by m.

From each base solution (x0, y0), we generate further solutions using
Brahmagupta-Pell composition with the fundamental solution.
We also need the conjugate chain from (x0, -y0).
"""

from __future__ import annotations

from math import isqrt


def is_square(n):
    """Check if n is a perfect square."""
    if n < 0:
        return False
    root = isqrt(n)
    return root * root == n


def factorize(n):
    """Return prime factorization as dict {prime: exponent}."""
    factors = {}
    d = 2
    while d * d <= n:
        while n % d == 0:
            factors[d] = factors.get(d, 0) + 1
            n //= d
        d += 1
    if n > 1:
        factors[n] = factors.get(n, 0) + 1
    return factors


def transform_exponents(prime_factors, f):
    """Transform exponents using function f."""
    result = 1
    for p, e in prime_factors.items():
        result *= p ** f(e)
    return result


def solve():
    """Solve Problem 261."""
    N = 10 ** 10
    L = isqrt(N // 2)

    pivots = set()

    for m in range(1, L + 1):
        D = m * (m + 1)

        # Compute sD: squarefree part of D
        factors_m = factorize(m)
        factors_m1 = factorize(m + 1)
        combined = dict(factors_m)
        for p, e in factors_m1.items():
            combined[p] = combined.get(p, 0) + e
        sD = transform_exponents(combined, lambda e: e % 2)

        # sm: product of primes in m raised to ceil(exponent/2)
        sm = transform_exponents(factors_m, lambda e: (e + 1) // 2)

        # Find base solutions: y from 0 to m, step sm
        base_sols = []
        for y in range(0, m + 1, sm):
            res = m + y * y
            if res % sD == 0 and is_square(res // sD):
                x = sD * isqrt(D // sD) * isqrt(res // sD)
                base_sols.append((x, y))

        # Fundamental solution to x^2 - D*y^2 = 1 is (2m+1, 2)
        xf = 2 * m + 1
        yf = 2

        def check_and_add(x, y):
            if x % m == 0 and (x // m - m - 1) % 2 == 0 and (y + m) % 2 == 0:
                n_val = (x // m - m - 1) // 2
                k = (y + m) // 2
                if n_val >= k:
                    pivots.add(k)

        # Generate solutions from each base solution
        # Both the direct chain and the conjugate chain
        for x0, y0 in base_sols:
            # Direct chain: (x0, y0) composed with (xf, yf)^n
            x, y = x0, y0
            while True:
                if y + m > 2 * N:
                    break
                check_and_add(x, y)
                x, y = xf * x + D * yf * y, xf * y + yf * x

            # Conjugate chain: start from (x0, -y0), compose with (xf, yf)
            # (x0, -y0) * (xf, yf) = (xf*x0 - D*yf*y0, yf*x0 - xf*y0)
            if y0 > 0:
                x, y = xf * x0 - D * yf * y0, yf * x0 - xf * y0
                # y might be negative initially; keep composing until y >= 0
                while y < 0:
                    x, y = xf * x + D * yf * y, xf * y + yf * x
                while True:
                    if y + m > 2 * N:
                        break
                    check_and_add(x, y)
                    x, y = xf * x + D * yf * y, xf * y + yf * x

    return sum(pivots)


def main():
    """Main entry point."""
    result = solve()
    print(result)


if __name__ == "__main__":
    main()

File: c/261.c
/*
 * Project Euler 261: Pivotal Square Sums
 *
 * Pell equation approach: for each m, solve x^2 - D*y^2 = m^2*(m+1)
 * with D = m*(m+1). Generate solutions via Brahmagupta-Pell composition.
 * Uses __int128 for large intermediate values.
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include <string.h>

typedef __int128 i128;

static int is_square_64(int64_t n) {
    if (n < 0) return 0;
    int64_t r = (int64_t)sqrt((double)n);
    /* Adjust for floating point imprecision */
    while (r > 0 && r * r > n) r--;
    while ((r + 1) * (r + 1) <= n) r++;
    return r * r == n;
}

static int64_t isqrt64(int64_t n) {
    if (n <= 0) return 0;
    int64_t r = (int64_t)sqrt((double)n);
    while (r > 0 && r * r > n) r--;
    while ((r + 1) * (r + 1) <= n) r++;
    return r;
}

/* Factorize n, store prime factors and exponents */
#define MAX_FACTORS 20
typedef struct {
    int64_t primes[MAX_FACTORS];
    int exps[MAX_FACTORS];
    int count;
} Factorization;

static void factorize(int64_t n, Factorization *f) {
    f->count = 0;
    int64_t d = 2;
    while (d * d <= n) {
        if (n % d == 0) {
            f->primes[f->count] = d;
            f->exps[f->count] = 0;
            while (n % d == 0) {
                f->exps[f->count]++;
                n /= d;
            }
            f->count++;
        }
        d++;
    }
    if (n > 1) {
        f->primes[f->count] = n;
        f->exps[f->count] = 1;
        f->count++;
    }
}

static int64_t ipow(int64_t base, int exp) {
    int64_t result = 1;
    for (int i = 0; i < exp; i++) result *= base;
    return result;
}

/* Pivot set using sorted array + binary search for dedup */
#define MAX_PIVOTS 200000
static int64_t pivots[MAX_PIVOTS];
static int num_pivots = 0;

static int cmp64(const void *a, const void *b) {
    int64_t va = *(const int64_t *)a;
    int64_t vb = *(const int64_t *)b;
    if (va < vb) return -1;
    if (va > vb) return 1;
    return 0;
}

static void add_pivot(int64_t k) {
    if (num_pivots < MAX_PIVOTS)
        pivots[num_pivots++] = k;
}

int main(void) {
    int64_t N = 10000000000LL; /* 10^10 */
    int64_t L = isqrt64(N / 2);

    for (int64_t m = 1; m <= L; m++) {
        i128 D = (i128)m * (m + 1);

        /* Factorize m and m+1 */
        Factorization fm, fm1;
        factorize(m, &fm);
        factorize(m + 1, &fm1);

        /* Compute sD (squarefree part of D) and sqrt(D/sD) */
        /* Combined factorization of m*(m+1) */
        int64_t combined_p[40];
        int combined_e[40];
        int combined_count = 0;

        for (int i = 0; i < fm.count; i++) {
            combined_p[combined_count] = fm.primes[i];
            combined_e[combined_count] = fm.exps[i];
            combined_count++;
        }
        for (int i = 0; i < fm1.count; i++) {
            int found = 0;
            for (int j = 0; j < combined_count; j++) {
                if (combined_p[j] == fm1.primes[i]) {
                    combined_e[j] += fm1.exps[i];
                    found = 1;
                    break;
                }
            }
            if (!found) {
                combined_p[combined_count] = fm1.primes[i];
                combined_e[combined_count] = fm1.exps[i];
                combined_count++;
            }
        }

        /* sD = product of p^(e%2) */
        int64_t sD = 1;
        for (int i = 0; i < combined_count; i++)
            if (combined_e[i] % 2 == 1)
                sD *= combined_p[i];

        /* sqrt(D/sD) */
        int64_t sqrt_D_over_sD = 1;
        for (int i = 0; i < combined_count; i++)
            sqrt_D_over_sD *= ipow(combined_p[i], combined_e[i] / 2);

        /* sm = product of p^ceil(e/2) for factors of m */
        int64_t sm = 1;
        for (int i = 0; i < fm.count; i++)
            sm *= ipow(fm.primes[i], (fm.exps[i] + 1) / 2);

        /* Find base solutions: y from 0 to m, step sm */
        i128 base_x[200], base_y[200];
        int num_base = 0;

        for (int64_t y = 0; y <= m; y += sm) {
            int64_t res = m + y * y;
            if (res % sD != 0) continue;
            int64_t quotient = res / sD;
            if (!is_square_64(quotient)) continue;
            i128 x = (i128)sD * sqrt_D_over_sD * isqrt64(quotient);
            if (num_base < 200) {
                base_x[num_base] = x;
                base_y[num_base] = y;
                num_base++;
            }
        }

        /* Fundamental solution: (2m+1, 2) */
        i128 xf = 2 * m + 1;
        i128 yf = 2;

        for (int b = 0; b < num_base; b++) {
            /* Direct chain */
            i128 x = base_x[b], y = base_y[b];
            while (1) {
                if (y + m > 2 * N) break;
                /* Check and add */
                if (x % m == 0) {
                    i128 val = x / m - m - 1;
                    if (val >= 0 && val % 2 == 0) {
                        i128 ym = y + m;
                        if (ym >= 0 && ym % 2 == 0) {
                            i128 n_val = val / 2;
                            i128 k = ym / 2;
                            if (n_val >= k && k > 0 && k <= N) {
                                add_pivot((int64_t)k);
                            }
                        }
                    }
                }
                i128 nx = xf * x + D * yf * y;
                i128 ny = xf * y + yf * x;
                x = nx; y = ny;
            }

            /* Conjugate chain */
            if (base_y[b] > 0) {
                x = xf * base_x[b] - D * yf * base_y[b];
                y = yf * base_x[b] - xf * base_y[b];
                while (y < 0) {
                    i128 nx = xf * x + D * yf * y;
                    i128 ny = xf * y + yf * x;
                    x = nx; y = ny;
                }
                while (1) {
                    if (y + m > 2 * N) break;
                    if (x % m == 0) {
                        i128 val = x / m - m - 1;
                        if (val >= 0 && val % 2 == 0) {
                            i128 ym = y + m;
                            if (ym >= 0 && ym % 2 == 0) {
                                i128 n_val = val / 2;
                                i128 k = ym / 2;
                                if (n_val >= k && k > 0 && k <= N) {
                                    add_pivot((int64_t)k);
                                }
                            }
                        }
                    }
                    i128 nx = xf * x + D * yf * y;
                    i128 ny = xf * y + yf * x;
                    x = nx; y = ny;
                }
            }
        }
    }

    /* Deduplicate and sum */
    qsort(pivots, num_pivots, sizeof(int64_t), cmp64);
    int64_t ans = 0;
    for (int i = 0; i < num_pivots; i++) {
        if (i == 0 || pivots[i] != pivots[i-1])
            ans += pivots[i];
    }
    printf("%lld\n", (long long)ans);
    return 0;
}
